### ISSUE TITLE
Alignment option for the last line of a paragraph

### ISSUE THREAD

**Issue by ghost** [typst/typst#476]
Currently, the last line of a paragraph is left-aligned.
For some situations one of the following results is wanted (all can be seen in books):
The last line:
1. should be left aligned (like now)
2. should be [completely filled](https://tex.stackexchange.com/a/497596), like all other lines
3. should be right-aligned if the last line is less than some percentage (often seen with a hyphenated word, but also without)
4. should be centered (all other lines are completely filled, like now)

Number 1,3,4 can already partly be achieved using `align(alignment, content)`.
This will always align using the given alignment, independent of the remaining whitespace.

The situation number 2. is probably not so easy to achieve and requires some thinking about performance impact.
One slow solution would be to calculate the paragraph like it is currently, then look at the last line and measure the whitespace.
After this, calculate a new word-spacing via arithmetic mean, so that the space at the end is exactly 0 now (new word-spacing is now >= old word-spacing). Then set the paragraph with this new word-spacing.
This will not perfectly work because new line breaks and hyphenations will appear, but probably nearly as wanted.
Then the last word could be right-aligned. Maybe, a much better algorithm exists.
When the last line is filled by less than half: the word space can be decreased, else increased.
It [seems](https://tex.stackexchange.com/a/41428), that this slow algorithm is exactly the way to achieve this.
The algorithms could also be called recursively where each recursion improves the result and increases the runtime. But probably one iteration is enough for most.

The following parameter could be added to `par` to achieve one of the 4 results:
```
last-line: alignment or none or function that takes the remaining space (absolute and relative?) and returns an alignment or none;
// default: alignment left.
```

If `none` is set or returned by the function, there is no alignment and the line is completely filled.

or (more general and easier to extend later)
```
last-line: (
  alignment: (then the same as above)
)
```
or
```
alignment: (
  last-line: (then the same as above)
)
```

Another solution would be to extend the general `align` to allow passing the parameters `function(absolute_used_space, relative_used_space) -> alignment | none` and `none` where `none` would use the whole text width.

**Comment by laurmaedje** [comment]
The last line isn't left aligned, but aligned as set with `set align(..)`. It can also be right-aligned. But this is of course still less flexible than what you propose.

**Comment by w1th0utnam3** [comment]
If something like this becomes a feature, it should also not be applied if the paragraph does not fill at least one line. As a use case I can imagine figure captions: 
 - If the caption has less than one full line it should be aligned with the surrounding `align`, e.g. centered.
 - If the caption has more than one full line it should be justified and the last line aligned with the proposed `last-line` alignment option, e.g. aligned with the start of the line.
 
 I don't think this can be achieved currently?

**Comment by laurmaedje** [comment]
@w1th0utnam3 I think what you want can be achieved in two ways at the moment:

- Option A: Make the caption left aligned _within_ an automatically sized block, but center align the block. Due to the block being auto-sized, it might fill exactly 100% of the page width. In that case, defer to option B.
- Option B: Do something with `layout` and `measure` to have separate logic for one line vs multiple lines.

**Comment by w1th0utnam3** [comment]
@laurmaedje Thanks for the suggestions! I guess with Option A you meant that it might *not* fill exactly 100%? In any case this is what happened when I tried it: the caption was slightly smaller than the document body depending on how it decided to do the line breaks. Option B seems to work though - although it doesn't feel super elegant. But it's nice that it works.

**Comment by laurmaedje** [comment]
Yeah, sometimes I lose a "not" when typing on the phone quickly. Sorry for the confusion.
